src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -5612,8 +5656,19 @@ void onFedAuthInfo(SqlFedAuthInfo fedAuthInfo, TDSTokenHandler tdsTokenHandler)
         SQLServerAccessTokenCallback callback = (SQLServerAccessTokenCallback) activeConnectionProperties
                 .get(SQLServerDriverObjectProperty.ACCESS_TOKEN_CALLBACK.toString());
 
-        if (authenticationString.equals(SqlAuthentication.NOT_SPECIFIED.toString()) && null != callback) {
-            fedAuthToken = callback.getAccessToken(fedAuthInfo.spn, fedAuthInfo.stsurl);
+        if (authenticationString.equals(SqlAuthentication.NOT_SPECIFIED.toString()) && null != accessTokenCallbackClass
+                && !accessTokenCallbackClass.isEmpty()) {
+            try {
+                Object[] msgArgs = {"accessTokenCallbackClass",
+                        "com.microsoft.sqlserver.jdbc.SQLServerAccessTokenCallback"};
+                SQLServerAccessTokenCallback callbackInstance = Util.newInstance(SQLServerAccessTokenCallback.class,
+                        accessTokenCallbackClass, null, msgArgs);
+                fedAuthToken = callbackInstance.getAccessToken(fedAuthInfo.stsurl, fedAuthInfo.spn);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        } else if (authenticationString.equals(SqlAuthentication.NOT_SPECIFIED.toString()) && null != callback) {
+            fedAuthToken = callback.getAccessToken(fedAuthInfo.stsurl, fedAuthInfo.spn);
```

Comment:
wouldn't this always be the case? we checked for this before and would've thrown an error if this was true right?

Reply:
It's possible if a user sets `authentication=ActiveDirectoryPassword` and sets the callback, the check we did before wouldn't fail because property values are appropriate. So when we get here, we wouldn't know the if the authenticationString is really `notSpecified`.